From e5d79a3888747e331066d332286f5872b908fc69 Mon Sep 17 00:00:00 2001
From: inogenous <123803852+inogenous@users.noreply.github.com>
Date: Sat, 9 Sep 2023 06:57:07 +0200
Subject: [PATCH] Bugfix: Prevent segfault when scrolling options

* Prevents segfault when scrolling options when "Centered menu scrolling" is set to `false`.
* The issue before was:
	* `curr_line_visible` is supposed to contain the index of the currently selected options item.
	* However, it was previously assigned to the value of `visible_items.size()` **after** an item had been added to `visible_items`
	* This made `curr_line_visible` be 1-indexed (counting starts at 1). This is the root cause of the problem.
	* The value of `curr_line_visible` is passed into `calcStartPos` as the parameter named `iCurrentLine`.
	* `calcStartPos` assumes that `iCurrentLine` is 0-indexed (counting starts at 0).
* With this change, it now possible to scroll the options when it displays more lines than can fit on screen and having "Centered menu scrolling" is set to `false`.
---
 src/options.cpp | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/options.cpp b/src/options.cpp
index d560b6381961f..38cce516868de 100644
--- a/src/options.cpp
+++ b/src/options.cpp
@@ -3475,10 +3475,10 @@ std::string options_manager::show( bool ingame, const bool world_options_only, b
         };
         for( int i = 0; i < static_cast<int>( page_items.size() ); i++ ) {
             if( is_visible( i ) ) {
-                visible_items.push_back( i );
                 if( i == iCurrentLine ) {
                     curr_line_visible = static_cast<int>( visible_items.size() );
                 }
+                visible_items.push_back( i );
             }
         }
 
